### ISSUE TITLE
unused: lint unused exported members in the `./internal` directory

### ISSUE THREAD

**Issue by qdm12** [dominikh/go-tools#1406]
Since Go code in the `./internal` cannot be imported by other projects, it would be nice to lint project-locally-unused Go types/variables/constants located in the `./internal` directory.

**Comment by dominikh** [comment]
In terms of implementation difficulty, this is identical to detecting any kind of unused exported object. The proposed idea is a filter to only flag those that are in `internal` packages, not all checked packages.

There's no concrete timeline for supporting finding unused exported objects, although the work on it has started.

**Comment by qdm12** [comment]
> In terms of implementation difficulty, this is identical to detecting any kind of unused exported object.

I suppose this is due to the checks crossing packages, whereas for unexported objects it's only package local right?

Would you think now would be a good time to start PR-ing for this feature, or do you have already opened PRs / local work waiting on your end?

Thanks!

**Comment by dominikh** [comment]
> I suppose this is due to the checks crossing packages, whereas for unexported objects it's only package local right?

Crossing them in the wrong direction, at that. Normal analyzes in Staticcheck (and the `go/analysis` framework) propagate information from dependencies to dependents. But figuring out which objects are used needs information to flow in the opposite direction.

But that's not the only difficulty. There are also a lot of edge cases involving interfaces and reflection. A method might only be used dynamically, for example.

> Would you think now would be a good time to start PR-ing for this feature, or do you have already opened PRs / local work waiting on your end?

There's some existing work already commited/pushed but not exposed, because it is incomplete. Generally speaking, however, I would advise against sending PRs. I usually take even longer to review PRs than I take implementing things myself :grimacing: This task in particular also requires a lot of intimate knowledge of how `unused` works, and unfortunately most of that is only recorded in my mind.

The issue tracker has a fair amount of prior discussions on the matter of whole-program unusedness.

**Comment by qdm12** [comment]
Thank you for taking the time explaining all this 👍

> I usually take even longer to review PRs than I take implementing things myself

Oh I feel you 😄 The ugly side of pull requests I guess 😸

> The issue tracker has a fair amount of prior discussions on the matter of whole-program unusedness.

Feel free to close mine if you feel it's duplicating another issue or could be bundled in another issue.

> But that's not the only difficulty. There are also a lot of edge cases involving interfaces and reflection. A method might only be used dynamically, for example.

Wouldn't it be worth it to have an imperfect linter and use for example golangci-lint's `//nolint:unused` on such niche cases? 🤔

**Comment by dominikh** [comment]
> Feel free to close mine if you feel it's duplicating another issue or could be bundled in another issue.

I'm leaving this issue open because it documents an interesting UI middle-ground between flagging all or no exported objects.

> Wouldn't it be worth it to have an imperfect linter and use for example golangci-lint's //nolint:unused on such niche cases?

The guiding principle for Staticcheck is to err on the side of false negatives, not false positives. This is especially true for a check as impactful as this one, which tells you to _delete your code_ and where it might not be obvious that the suggestion is wrong — removing code required for dynamic calls will cause failures at run time, not at compile time.

**Comment by alingse** [comment]
+  we need this to delete old codes.

**Comment by qdm12** [comment]
Any updates on this?

Could we at least have something linting only unused exported pure functions and types in the `internal` directory, but not methods? I don't think one could use them dynamically right?

**Comment by hugosjoberg** [comment]
I made a script because I wanted to delete a bunch of internal packages which I no longer used. Run this script, delete the packages and run it again until it doesn't spit anything out.
```sh
# List all packages in the project
go list ./... > all_packages.txt

# List only packages that are actually imported by others (excluding main packages)
go list -f '{{join .Imports "\n"}}' ./... | grep '^github.com/amberdata/market-data-pipeline/src/redpanda-connect-plugin-v2/internal/' | sort | uniq > imported_internal_packages.txt

# Find packages in 'internal' that aren't used by other packages
grep 'internal/' all_packages.txt | grep -v -F -f imported_internal_packages.txt
```
